core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+							for i in 0..<x_thumbnail*y_thumbnail {
+								pixel := palette[old_pixels[i]]
+								pixels[i] = pixel.r
+								pixels[i + 1] = pixel.g
+								pixels[i + 2] = pixel.b
+							}
+
+							info: ^image.JPEG_Info
+							if img.metadata == nil {
+								info = new(image.JPEG_Info)
+							} else {
+								info = img.metadata.(^image.JPEG_Info)
+							}
+							info.jfxx_app0 = image.JFXX_APP0{
+								extension_code,
+								x_thumbnail,
+								y_thumbnail,
+								pixels,
+							}
+							img.metadata = info
+						}
+					case:
+						return img, .Invalid_JFXX_Extension_Code
+					}
+				} else {
+					fmt.printfln("Unrecognized APP0 identifier \"%s\". Skipping", string(ident[:]))
+					// - 1 for the NUL byte
+					compress.read_slice(ctx, length - len(ident) - 1) or_return
+					continue
+				}
+			// case .APP1: // Exif metadata
+				// unimplemented("APP1")
+			case .COM:
+				length := (compress.read_data(ctx, u16be) or_return) - 2
+				comment := string(compress.read_slice(ctx, cast(int)length) or_return)
+				if .return_metadata in options {
+					if info, ok := img.metadata.(^image.JPEG_Info); ok {
+						if info.comments == nil {
+							info.comments = make([dynamic]string, 0, 8, allocator)
+						}
+						append(&info.comments, strings.clone(comment))
+					}
+				}
+			case .DQT:
+				length := cast(int)(compress.read_data(ctx, u16be) or_return) - 2
+
+				for length > 0 {
+					precision_and_index := compress.read_u8(ctx) or_return
+					precision := precision_and_index >> 4
+					index := precision_and_index & 0xF
+
+					if precision != 0 && precision != 1 {
+						return img, .Invalid_Quantization_Table_Precision
+					}
+
+					if index < 0 || index > 3 {
+						return img, .Invalid_Quantization_Table_Index
+					}
+
+					// When precision is 0, we read 64 u8s.
+					// when it's 1, we read 64 u16s.
+					table_bytes := 64
+					if precision == 1 {
+						table_bytes = 128
+						table := compress.read_slice(ctx, table_bytes) or_return
+						for v, i in slice.reinterpret([]u16be, table) {
+							quantization[index][i] = v
+						}
+					} else {
+						table := compress.read_slice(ctx, table_bytes) or_return
+						for v, i in table {
+							quantization[index][i] = cast(u16be)v
+						}
+					}
+
+					length -= table_bytes + 1
+				}
+			case .DHT:
+				length := (compress.read_data(ctx, u16be) or_return) - 2
+
+				for length > 0 {
+					type_index := compress.read_u8(ctx) or_return
+					type := cast(Coefficient)((type_index >> 4) & 0xF)
+					index := type_index & 0xF
+
+					if type != .DC && type != .AC {
+						return img, .Invalid_Huffman_Coefficient_Type
+					}
+
+					if index < 0 || index > 3 {
+						return img, .Invalid_Huffman_Table_Index
+					}
+
+					lengths := compress.read_slice(ctx, HUFFMAN_MAX_BITS) or_return
+					num_symbols := 0
+					for length, i in lengths {
+						num_symbols += cast(int)length
+						huffman[type][index].offsets[i + 1] = cast(u8)num_symbols
+					}
```

Comment:
Malicious JPEG can provide lengths that cause `num_symbols` to exceed `HUFFMAN_MAX_SYMBOLS` which can lead to later OOB. Should add bounds checking to `num_symbols` here.